spec.html

```diff
@@ -87,80 +48,63 @@ <h1>Uint8Array.prototype.toHex ( )</h1>
+    1. Let _strict_ be ToBoolean(? Get(_opts_, *"strict"*)).
+    1. NOTE: The order of validation and decoding in the algorithm below is not observable. Implementations are encouraged to perform them in whatever order is most efficient, possibly interleaving validation and stripping whitespace with decoding, as long as the behaviour is observably equivalent.
+    1. Let _input_ be StringToCodePoints(_string_).
+    1. If _alphabet_ is *"base64url"*, then
+      1. If _input_ contains U+002B (PLUS SIGN) or U+002F (SOLIDUS), throw a *SyntaxError* exception.
+      1. Replace all occurrences of U+002D (HYPHEN-MINUS) in _input_ with U+002B (PLUS SIGN).
+      1. Replace all occurrences of U+005F (LOW LINE) in _input_ with U+002F (SOLIDUS).
+    1. NOTE: When _strict_ is *false*, the algorithm below is equivalent to the <a href="https://infra.spec.whatwg.org/#forgiving-base64-decode">forgiving-base64 decode</a> operation in HTML.
+    1. If _strict_ is *false*, then
+      1. Remove all occurrences of U+0009 (TAB), U+000A (LF), U+000C (FF), U+000D (CR), and U+0020 (SPACE) from _input_.
+    1. Let _inputLength_ be the length of _input_.
+    1. If _inputLength_ modulo 4 is 0, then
+      1. If _input_ is not empty and the last element of _input_ is U+003D (EQUALS SIGN), then
+        1. Remove the last element of _input_.
+        1. Set _inputLength_ to _inputLength_ - 1.
+        1. If _input_ is not empty and the last element of _input_ is U+003D (EQUALS SIGN), then
+          1. Remove the last element of _input_.
+          1. Set _inputLength_ to _inputLength_ - 1.
     1. Else,
-      1. Assert: _alphabet_ is *"base64url*".
-      1. If _characters_ cannot result from applying the base64url encoding specified in section 5 of RFC 4648 to some sequence of bytes, throw a *SyntaxError* exception.
-      1. Let _bytes_ be the unique sequence of bytes such that applying the base64url encoding specified in section 5 of RFC 4648 to that sequence would produce _characters_.
-    1. Let _resultLength_ be the number of bytes in _bytes_.
-    1. Let _result_ be ? AllocateTypedArray(*"Uint8Array"*, %Uint8Array%, %Uint8Array.prototype%, _resultLength_).
+      1. If _strict_ is *true*, throw a *SyntaxError* exception.
+    1. If _input_ contains any elements which are not also elements of the standard base64 alphabet, throw a *SyntaxError* exception.
+    1. Let _lastChunkSize_ be _inputLength_ modulo 4.
+    1. If _lastChunkSize_ is 1, then
+      1. Throw a *SyntaxError* exception.
+    1. Else if _lastChunkSize_ is 2, then
+      1. Append U+0041 (LATIN CAPITAL LETTER A) to _input_ twice.
+    1. Else if _lastChunkSize_ is 3, then
+      1. Append U+0041 (LATIN CAPITAL LETTER A) to _input_.
+    1. Let _bytes_ be the unique (possibly empty) sequence of bytes such that applying the base64 encoding specified in section 4 of RFC 4648 to that sequence would produce _input_.
```

Comment:
Revisiting this step, solving a puzzle seems a bit too opaque for what is fundamentally a simple algorithm. Maybe it would be better to specify things more directly?
```
    1. Let _inputLength_ be the length of _input_.
    1. If _inputLength_ modulo 4 is 0, then
      1. NOTE: If _input_ is valid, it might end with one or two padding characters that must be removed before comparing its contents against the standard base64 alphabet.
      1. If _input_ is not empty and the last element of _input_ is U+003D (EQUALS SIGN), remove the last element of _input_.
      1. If _input_ is not empty and the last element of _input_ is U+003D (EQUALS SIGN), remove the last element of _input_.
      1. Set _inputLength_ to the length of _input.
    1. Else,
      1. If _strict_ is *true*, throw a *SyntaxError* exception.
    1. If _input_ contains any elements which are not also elements of the standard base64 alphabet, throw a *SyntaxError* exception.
    1. Let _lastChunkSize_ be _inputLength_ modulo 4.
    1. If _lastChunkSize_ is 1, then
      1. Throw a *SyntaxError* exception.
    1. Else if _lastChunkSize_ is 2, then
      1. Append U+0041 (LATIN CAPITAL LETTER A) to _input_ twice.
    1. Else if _lastChunkSize_ is 3, then
      1. Append U+0041 (LATIN CAPITAL LETTER A) to _input_.
    1. Let _paddedInputLength_ be the length of _input_.
    1. Let _bytes_ be an empty List of octets.
    1. Let _buffer_ be 0.
    1. Let _i_ be 0.
    1. Repeat, while _i_ &lt; _paddedInputLength_,
      1. Set _buffer_ to (_buffer_ × 64) + StringIndexOf(CodePointsToString(standard base64 alphabet), _input_[_i_], 0).
      1. If _i_ modulo 4 is 3, then
        1. Append floor(_buffer_ / 256<sup>2</sup>) to _bytes_.
        1. Append floor((_buffer_ modulo 256<sup>2</sup>) / 256) to _bytes_.
        1. Append (_buffer_ modulo 256) to _bytes_.
        1. Set _buffer_ to 0.
      1. Set _i_ to _i_ + 1.
    1. Let _byteLength_ be the length of _bytes_.
    1. If _lastChunkSize_ is 2, then
      1. If _strict_ is *true* and _bytes_[_byteLength_ - 2] is not 0, throw a *SyntaxError* exception.
      1. Set _byteLength_ to _byteLength_ - 2.
    1. Else if _lastChunkSize_ is 3, then
      1. If _strict_ is *true* and _bytes_[_byteLength_ - 1] is not 0, throw a *SyntaxError* exception.
      1. Set _byteLength_ to _byteLength_ - 1.
    1. Let _result_ be ? AllocateTypedArray(*"Uint8Array"*, %Uint8Array%, %Uint8Array.prototype%, _byteLength_).
    1. Set the value at each index of _result_.[[ViewedArrayBuffer]].[[ArrayBufferData]] to the value at the corresponding index of _bytes_.
    1. Return _result_.
```

Reply:
That seems much more opaque to me. What we mean is just do bog-standard base64 decoding. We shouldn't make implementations figure that that we've written down the standard base64 decoding algorithm, which they'll then ignore and use their own existing base64 decoding logic; we should just tell them to do base64 decoding.

The only reason it's phrased somewhat awkwardly is because the RFC doesn't actually specify a decoding algorithm, only encoding. But despite that awkwardness the current text seems much better to me than an explicit algorithm. The world does not need the base64 decoding algorithm written out yet again.

Reply:
> The world does not need the base64 decoding algorithm written out yet again.

Perhaps not, but "the \<value> such that applying \<_encoding_ algorithm> would produce _input_" is hardly a good reference to any prior example. And from another perspective, ECMA-262 explicitly specifies all kinds of common specialized algorithms, including near-trivial ones like [UTF16EncodeCodePoint](https://tc39.es/ecma262/multipage/ecmascript-language-source-code.html#sec-utf16encodecodepoint) and longer ones like [Encode](https://tc39.es/ecma262/multipage/global-object.html#sec-encode), and also a number of non-specialized algorithms such as [StringIndexOf](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-stringindexof) and [StringPad](https://tc39.es/ecma262/multipage/text-processing.html#sec-stringpad)—so what makes base64 special?

Reply:
> Perhaps not, but "the <value> such that applying <_encoding_ algorithm> would produce _input_" is hardly a good reference to any prior example.

... Isn't it? I mean, I'm open to other wordings for the thing I'm getting at but I don't really think there's much possibility for confusion with that wording. I am not going to write down the base64 algorithm again.

> so what makes base64 special?

None of those algorithms are widely implemented under a common name, specified elsewhere in a document easy to incorporate by reference, and unambiguous. This one is.

Reply:
Can this one be made into a link?

Reply:
> None of those algorithms are widely implemented under a common name, specified elsewhere in a document easy to incorporate by reference, and unambiguous. This one is.

Actually, many of them are. For example, UTF16EncodeCodePoint is [RFC 2781 section 2.1](https://www.rfc-editor.org/rfc/rfc2781#section-2.1) and encodeURIComponent is the WHATWG URL Standard [UTF-8 percent-encode](https://url.spec.whatwg.org/#utf-8-percent-encode) … using the [component percent-encode set].

Reply:
I'd be OK swapping out our Encode with a reference to WHATWG. But I would not exactly describe it as _widely implemented_, at least not nearly as much as base64.

Reply:
Conceded. I still really dislike algorithm steps that are basically "solve this puzzle" or "find a preimage", but base64 decoding is in fact well understood so in practice implementations will "_figure that that we've ~~written down~~ **obliquely referenced** the standard base64 decoding algorithm **[implied by RFC 4648 but not written therein]**, which they'll then ignore and use their own existing base64 decoding logic_". To that end, one final suggestion:
```suggestion
    1. Let _bytes_ be the unique (possibly empty) sequence of bytes resulting from decoding _input_ as base64 (such that applying the base64 encoding specified in section 4 of RFC 4648 to _bytes_ would produce _input_).
```

Reply:
SGTM, done.